clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -538,14 +538,86 @@ check_duplicate_syscall_with_same_pc()
     return true;
 }
 
+bool
+check_rseq_side_exit_discontinuity()
+{
+#if defined(X86_64) || defined(X86_32) || defined(ARM_64)
+    constexpr addr_t ADDR_ONE = 0xeba4ad4;
+    constexpr addr_t ADDR_TWO = 0xeba4ae4;
+    // Negative test: branches that do not go to their targets due to rseq side
+    // exit.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_marker(1, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
+#    if defined(X86_64) || defined(X86_32)
+            // eba4ad4:     74 32        je     11a9ff29
+            gen_branch_encoded(1, ADDR_ONE, { 0x74, 0x32 }),
+            // eba4ad8:     0f 11 40 10         movups %xmm0,0x10(%rax)
+            gen_instr_encoded(ADDR_ONE + 2, { 0x0f, 0x11, 0x40, 0x10 }),
+            // Memref for the above memory-accessing instr.
+            gen_data(1, false, ADDR_ONE + 2, 4),
+            // eba4ae4:     24 01        and    $0x1,%al
+            gen_instr_encoded(ADDR_TWO, { 0x24, 0x01 })
+#    elif defined(ARM_64)
+            // eba4ad4:     540001a1     b.ne   71019df0
+            gen_branch_encoded(1, ADDR_ONE, 0x540001a1),
+            // eba4ad8:     b8206ac1     str    w1, [x22, x0]
+            gen_instr_encoded(ADDR_ONE + 4, 0xb8206ac1),
+            // Memref for the above memory-accessing instr.
+            gen_data(1, false, ADDR_ONE + 4, 4),
+            // eba4ae4:     92800013     mov    x19, #0x1
+            gen_instr_encoded(ADDR_TWO, 0x92800013),
```

Comment:
The number of hardcoded encodings is growing here: it is probably time to put in code to use DR's IR to construct these instructions and automate the encodings.

Reply:
Done. Added an IR based memref generation.